src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+            .invoke_proto_method(service_method_uri(), call_options, request_payload)
+            .await
+            .expect("invoking method should have succeeded");
+        // THEN the response contains the expected payload
+        assert_eq!(response.value, "Hello World");
+        assert!(!rpc_client.contains_pending_request(&message_id));
+    }
+
+    #[tokio::test]
+    async fn test_invoke_method_fails_with_remote_error() {
+        // GIVEN an RPC client
+        let listener_ref: Arc<Mutex<Option<Arc<dyn UListener>>>> = Arc::new(Mutex::new(None));
+        let response_listener = listener_ref.clone();
+
+        let mut mock_transport = MockTransport::default();
+        mock_transport.expect_do_register_listener().returning(
+            move |_source_filter, _sink_filter, listener| {
+                let mut v = listener_ref.lock().unwrap();
+                *v = Some(listener);
+                Ok(())
+            },
+        );
+        // and a remote service oration that returns an error
+        mock_transport
+            .expect_do_send()
+            .returning(move |request_message| {
+                let error = UStatus::fail_with_code(UCode::NOT_FOUND, "no such object");
+                let response_message = UMessageBuilder::response_for_request(
+                    request_message.attributes.as_ref().unwrap(),
+                )
+                .with_comm_status(UCode::NOT_FOUND)
+                .build_with_protobuf_payload(&error)
+                .unwrap();
+                let cloned_listener = response_listener.lock().unwrap().take().unwrap().clone();
+                tokio::spawn(async move { cloned_listener.on_receive(response_message).await });
+                Ok(())
+            });
+
+        let client = InMemoryRpcClient::new(Arc::new(mock_transport), new_uri_provider())
+            .await
+            .unwrap();
+
+        // WHEN invoking the remote service operation
+        let message_id = UUID::build();
+        let mut call_options = CallOptions::default();
+        call_options.with_message_id(message_id.clone());
+        let response = client
+            .invoke_method(service_method_uri(), CallOptions::default(), None)
+            .await;
+
+        // THEN the invocation has failed with the error returned from the service
+        assert!(response.is_err_and(|e| { matches!(e, ServiceInvocationError::NotFound(_msg)) }));
+        assert!(!client.contains_pending_request(&message_id));
+    }
+
+    #[tokio::test]
+    async fn test_invoke_method_times_out() {
+        // GIVEN an RPC client
+        let mut mock_transport = MockTransport::default();
+        mock_transport
+            .expect_do_register_listener()
+            .returning(move |_source_filter, _sink_filter, _listener| Ok(()));
+        // and a remote service operation that does not return a response
+        mock_transport
+            .expect_do_send()
+            .returning(move |_request_message| Ok(()));
```

Comment:
Interesting. So in this case your mock transport nominally would succeed if not for the very small TTL of 20ms?

Am I reading this correctly?

Reply:
The client side transport sends out the request successfully, but there is no response message being received. So the request will time out, regardless of the concrete TTL value. I used a very small one, because I do not want to delay the execution of the test case unnecessarily ...